lib/autoload/condition.rb

```diff
@@ -0,0 +1,43 @@
+module Condition
+  # Finds all conditionals that depend on the given object.
+  #
+  # @param [Course::Assessment, Course::Achievement] dependent_object An assessment or
+  #   achievement that conditionals depends on
+  # @return [Object] acts_as_conditional Objects that depend on the condition_object
+  def self.conditionals_of(dependent_object)
+    condition_classes_of(dependent_object).map do |condition_name|
+      Course::Condition.find_by_sql(<<-SQL)
+        SELECT * FROM course_conditions cc
+          INNER JOIN course_condition_#{condition_name.demodulize.downcase.pluralize} ccs
+          ON cc.actable_type = '#{condition_name}'
+            AND cc.actable_id = ccs.id
+            AND ccs.#{dependent_object.class.name.demodulize.downcase}_id = #{dependent_object.id}
+        WHERE course_id = #{dependent_object.course_id}
+      SQL
+    end.flatten.map(&:conditional)
+  end
+
+  private
+
+  # Finds condition classes that depend on the dependent_object. For example, if the
+  # dependent_object is a Course::Achievement object, this method should return
+  # Course::Condition::Achievement.
+  def self.condition_classes_of(dependent_object)
+    dependent_class_to_condition_class_mapping.select do |dependent_name, _|
+      dependent_name == dependent_object.class.name
+    end.values
+  end
+
+  # Finds the mapping of dependent classes to condition classes. For example,
+  # {
+  #   'Course::Achievement' => 'Course::Condition::Achievement'
+  #   'Course::Assessment' => 'Course::Condition::Assessment'
+  # }
+  def self.dependent_class_to_condition_class_mapping
+    Course::Condition::ALL_CONDITIONS.map do |condition_name|
+      condition_name.constantize.dependent_classes.map do |dependent_name|
+        { dependent_name => condition_name }
```

Comment:
What if the dependent class is depended on by different types of conditions? The hash will only keep one of the dependent classes.

eg Level condition depends on Level. LevelAndAchievement depends on Level.